transport-classes-io_uring/src/main/java/io/netty/channel/uring/CompletionQueue.java

```diff
@@ -0,0 +1,160 @@
+    private final int ringMask;
+    private int ringHead;
+
+    CompletionQueue(long kHeadAddress, long kTailAddress, long kRingMaskAddress, long kRingEntriesAddress,
+                    long kOverflowAddress, long completionQueueArrayAddress, int ringSize, long ringAddress,
+                    int ringFd) {
+        this.kHeadAddress = kHeadAddress;
+        this.kTailAddress = kTailAddress;
+        this.completionQueueArrayAddress = completionQueueArrayAddress;
+        this.ringSize = ringSize;
+        this.ringAddress = ringAddress;
+        this.ringFd = ringFd;
+
+        ringEntries = PlatformDependent.getIntVolatile(kRingEntriesAddress);
+        ringMask = PlatformDependent.getIntVolatile(kRingMaskAddress);
+        ringHead = PlatformDependent.getIntVolatile(kHeadAddress);
+    }
+
+    /**
+     * Returns {@code true} if any completion event is ready to be processed by
+     * {@link #process(CompletionCallback)}, {@code false} otherwise.
+     */
+    boolean hasCompletions() {
+        return ringHead != PlatformDependent.getIntVolatile(kTailAddress);
+    }
+
+    @Override
+    public int getAsInt() {
+        return count();
+    }
+
+    int count() {
+        return PlatformDependent.getIntVolatile(kTailAddress) - ringHead;
+    }
+
+    /**
+     * Process the completion events in the {@link CompletionQueue} and return the number of processed
+     * events.
+     */
+    int process(CompletionCallback callback) {
+        int tail = PlatformDependent.getIntVolatile(kTailAddress);
+        int i = 0;
+        while (ringHead != tail) {
+            long cqeAddress = completionQueueArrayAddress + (ringHead & ringMask) * CQE_SIZE;
+
+            long udata = PlatformDependent.getLong(cqeAddress + CQE_USER_DATA_FIELD);
+            int res = PlatformDependent.getInt(cqeAddress + CQE_RES_FIELD);
+            int flags = PlatformDependent.getInt(cqeAddress + CQE_FLAGS_FIELD);
+
+            //Ensure that the kernel only sees the new value of the head index after the CQEs have been read.
+            ringHead++;
+            PlatformDependent.putIntOrdered(kHeadAddress, ringHead);
+
+            i++;
+
+            if (logger.isTraceEnabled()) {
+                logger.trace("completed(ring {}): {}(id={}, res={})",
+                        ringFd, Native.opToStr(UserData.decodeOp(udata)), UserData.decodeId(udata), res);
+            }
+            try {
+                decode(res, flags, udata, callback);
+            } catch (Error e) {
+                throw e;
+            } catch (Throwable throwable) {
+                handleLoopException(throwable);
+            }
+        }
+        return i;
+    }
+
+    private static void handleLoopException(Throwable throwable) {
+        logger.warn("Unexpected exception in the IO event loop.", throwable);
+
+        // Prevent possible consecutive immediate failures that lead to
+        // excessive CPU consumption.
+        try {
+            Thread.sleep(100);
```

Comment:
this was in the original code: is still valid?

Reply:
yes